### PR TITLE
Issue 249

### PR DESCRIPTION
**Pull request by planger**
#249

### CODE REVIEW

**Comment by planger** [comment]
Sorry about the whitespace changes due to regeneration. Do you want me to revert them? I guess they will re-appear on every re-generation though. Thanks!

File: plugins/org.eclipse.papyrus.uml.interaction.model.edit/src-gen/org/eclipse/papyrus/uml/interaction/model/provider/MExecutionItemProvider.java [hunk 1]
```diff
@@ -66,13 +66,13 @@ public MExecutionItemProvider(AdapterFactory adapterFactory) {
 	 * @generated
 	 */
 	protected void addStartPropertyDescriptor(Object object) {
-		itemPropertyDescriptors.add(createItemPropertyDescriptor(
```
Comment (cdamus) [inline plugins/org.eclipse.papyrus.uml.interaction.model.edit/src-gen/org/eclipse/papyrus/uml/interaction/model/provider/MExecutionItemProvider.java]: These formatting changes look like the "format code" property in "Templates and Merge" in the genmodel didn't work?  Which would be odd.  Maybe just explicitly format the `src-gen/` folder before committing?

File: plugins/org.eclipse.papyrus.uml.interaction.model.edit/src-gen/org/eclipse/papyrus/uml/interaction/model/provider/MExecutionItemProvider.java [hunk 2]
```diff
@@ -110,7 +110,7 @@ public Object getImage(Object object) {
 	@Override
 	public String getText(Object object) {
 		String label = ((MExecution) object).getName();
-		return (label == null) || (label.length() == 0) ? getString("_UI_MExecution_type") : //$NON-NLS-1$
```
Comment (cdamus) [inline plugins/org.eclipse.papyrus.uml.interaction.model.edit/src-gen/org/eclipse/papyrus/uml/interaction/model/provider/MExecutionItemProvider.java]: Again, not whitespace but formatting, which should not be changing with re-generation.

File: plugins/org.eclipse.papyrus.uml.interaction.model/model/seqd.ecore [hunk 3]
```diff
@@ -114,6 +114,20 @@
         </eAnnotations>
       </eParameters>
     </eOperations>
+    <eOperations name="getBottomMostElement" lowerBound="1">
```
Comment (cdamus) [inline plugins/org.eclipse.papyrus.uml.interaction.model/model/seqd.ecore]: Bottommost is one word (or, at least, "bottom-most") so the M should be lower case.

File: plugins/org.eclipse.papyrus.uml.interaction.model/model/seqd.ecore [hunk 4]
```diff
@@ -181,7 +195,7 @@
       </eParameters>
       <eParameters name="offset" lowerBound="1" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EInt"/>
       <eParameters name="height" lowerBound="1" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EInt"/>
-      <eParameters name="metaclass" lowerBound="1" eType="ecore:EClass ../../org.eclipse.emf.ecore/model/Ecore.ecore#//EClass"/>
```
Comment (cdamus) [inline plugins/org.eclipse.papyrus.uml.interaction.model/model/seqd.ecore]: Good catch.

File: plugins/org.eclipse.papyrus.uml.interaction.model/model/seqd.ecore [hunk 5]
```diff
@@ -114,6 +114,20 @@
         </eAnnotations>
       </eParameters>
     </eOperations>
+    <eOperations name="getBottomMostElement" lowerBound="1">
+      <eAnnotations source="http://www.eclipse.org/emf/2002/GenModel">
+        <details key="documentation" value="Returns the bottommost element, such as a message end, contained by this interaction."/>
```
Comment (cdamus) [inline plugins/org.eclipse.papyrus.uml.interaction.model/model/seqd.ecore]: Exactly!  😉

File: plugins/org.eclipse.papyrus.uml.interaction.model/src-gen/org/eclipse/papyrus/uml/interaction/internal/model/impl/MInteractionImpl.java [hunk 6]
```diff
@@ -337,4 +341,21 @@ MMessageImpl addMessage(MMessageImpl message) {
...
 	}
 
+	/**
+	 * <!-- begin-user-doc --> <!-- end-user-doc -->
+	 * 
+	 * @generated NOT
+	 */
+	@Override
+	public Optional<? extends MElement<?>> getBottomMostElement() {
+		Stream<MMessageEnd> messageEnds = getMessages().stream()
```
Comment (cdamus) [inline plugins/org.eclipse.papyrus.uml.interaction.model/src-gen/org/eclipse/papyrus/uml/interaction/internal/model/impl/MInteractionImpl.java]: The bottommost element could be an `MExecution`.

File: plugins/org.eclipse.papyrus.uml.interaction.model.edit/src-gen/org/eclipse/papyrus/uml/interaction/model/provider/MExecutionItemProvider.java (hunk 1 above)
Comment (planger) [reply plugins/org.eclipse.papyrus.uml.interaction.model.edit/src-gen/org/eclipse/papyrus/uml/interaction/model/provider/MExecutionItemProvider.java]: I actually tried formatting src-gen folder and it didn't lead to any changes -- and the version I committed is the result (with and without explicitly formatting the src-gen folders). Did we maybe commit the files without being explicitly formatted or without the format code property switched on initially?

File: plugins/org.eclipse.papyrus.uml.interaction.model.edit/src-gen/org/eclipse/papyrus/uml/interaction/model/provider/MExecutionItemProvider.java (hunk 2 above)
Comment (planger) [reply plugins/org.eclipse.papyrus.uml.interaction.model.edit/src-gen/org/eclipse/papyrus/uml/interaction/model/provider/MExecutionItemProvider.java]: Right, whitespace wasn't the right word. Formatting the src-gen folders didn't change anything (see previous comment).

File: plugins/org.eclipse.papyrus.uml.interaction.model/model/seqd.ecore (hunk 3 above)
Comment (planger) [reply plugins/org.eclipse.papyrus.uml.interaction.model/model/seqd.ecore]: Good point! I'll fix that in a subsequent commit.

**Comment by planger** [comment]
Thanks a lot for the review! I added a new commit to this branch to address the comment regarding the method rename and the support for MExecutions.

File: plugins/org.eclipse.papyrus.uml.interaction.model/src-gen/org/eclipse/papyrus/uml/interaction/internal/model/impl/MInteractionImpl.java [hunk 7]
```diff
@@ -347,15 +348,19 @@ MMessageImpl addMessage(MMessageImpl message) {
...
+	public Optional<? extends MElement<? extends Element>> getBottommostElement() {
 		Stream<MMessageEnd> messageEnds = getMessages().stream()
 				.flatMap((m) -> Stream.of(m.getSend(), m.getReceive())).filter(Optional::isPresent)
 				.map(Optional::get);
-		return messageEnds.max(Comparator.comparingInt(this::bottomOfMMessageEnd));
+		Stream<MOccurrence<?>> occurrances = getLifelines().stream()
+				.flatMap((lifeline) -> lifeline.getExecutions().stream())
+				.flatMap((e) -> Stream.of(e.getStart(), e.getFinish())).filter(Optional::isPresent)
+				.map(Optional::get);
+		return Stream.concat(messageEnds, occurrances).max(Comparator.comparingInt(this::bottomOfMElement));
```
Comment (cdamus) [inline plugins/org.eclipse.papyrus.uml.interaction.model/src-gen/org/eclipse/papyrus/uml/interaction/internal/model/impl/MInteractionImpl.java]: I think that once we add support for other things that cover lifelines (combined fragments, state invariants, interaction references) then the `MLifeline` API will need some union feature that simplifies this.

**Comment by planger** [comment]
Thanks for the review!

**Pull request merged**